streams/src/main/java/org/apache/kafka/streams/processor/internals/PunctuationSchedule.java

```diff
@@ -59,14 +59,19 @@ boolean isCancelled() {
         return isCancelled;
     }
 
-    public PunctuationSchedule next(long currTimestamp) {
-        PunctuationSchedule nextSchedule;
-        // we need to special handle the case when it is firstly triggered (i.e. the timestamp
-        // is equal to the interval) by reschedule based on the currTimestamp
-        if (timestamp == 0L)
-            nextSchedule = new PunctuationSchedule(value, currTimestamp + interval, interval, punctuator, cancellable);
-        else
-            nextSchedule = new PunctuationSchedule(value, timestamp + interval, interval, punctuator, cancellable);
+    public PunctuationSchedule next(final long currTimestamp) {
+        long nextPunctuationTime;
+        // avoid scheduling a new punctuations immediately, this can happen:
+        // - when using STREAM_TIME punctuation and there was a gap i.e., no data was
+        //   received for more than 2*interval (also happens on first STREAM_TIME punctuation i.e., when timestamp == 0L)
+        // - when using WALL_CLOCK_TIME and there was a gap i.e., punctuation was delayed for more than 2*interval (GC pause, overload, ...)
+        if (timestamp + interval < currTimestamp) {
+            nextPunctuationTime = currTimestamp + interval;
+        } else {
+            nextPunctuationTime = timestamp + interval;
+        }
```

Comment:
I see why I got confused. I wrongly interpreted the `if` branch as the "regular case" (no missed punctuation interval) but the `else` branch covers this case. Maybe it's worth to switch both branches?

Btw: I think the condition should be `timestamp + interval <= currTimestamp` -- otherwise, if the interval is 10, and next scheduled punctuation is 20, and the triggering record is 30, we schedule 30 next (instead of 40), trigger again, and than schedule 40.

We should extend the test to cover this case, too.